File: ios/Classes/lib/notifications/managers/ActionReceivedManager.swift
//
//  ActionReceivedManager.swift
//  awesome_notifications
//
//  Created by Sehii Moisa on 20/10/20.
//

import Foundation

public class ActionReceivedManager {
    
    static let shared:SharedManager = SharedManager(tag: Definitions.SHARED_ACTIONS)
    
    public static func removeAction(id:Int) -> Bool {
        return shared.remove(referenceKey: String(id));
    }

    public static func listActions() -> [ActionReceived] {
        var returnedList:[ActionReceived] = []
        let dataList = shared.getAllObjects()
        
        for data in dataList {
            let received:ActionReceived = ActionReceived(nil).fromMap(arguments: data) as! ActionReceived
            returnedList.append(received)
        }
        
        return returnedList
    }

    public static func saveAction(received:ActionReceived) {
        shared.set(received.toMap(), referenceKey: String(describing: received.id))
    }

    public static func getActionByKey(id:Int) -> ActionReceived? {
        return ActionReceived(nil).fromMap(arguments: shared.get(referenceKey: String(id))) as? ActionReceived
    }

    public static func cancelAllActions() {
        let receivedList = shared.getAllObjects();
        
        for received:[String:Any?] in receivedList {
            cancelAction(id: received["id"] as! Int);
        }
    }

    public static func cancelAction(id:Int) {
        _ = shared.remove(referenceKey: String(id))
    }
    
}

File: android/src/main/java/me/carda/awesome_notifications/notifications/managers/ActionReceivedManager.java
package me.carda.awesome_notifications.notifications.managers;

import android.content.Context;

import com.google.common.reflect.TypeToken;

import java.lang.reflect.Type;
import java.util.List;

import me.carda.awesome_notifications.Definitions;
import me.carda.awesome_notifications.notifications.models.returnedData.ActionReceived;

public class ActionReceivedManager {

    private static SharedManager<ActionReceived> shared = new SharedManager<>();
    private static Type typeToken = new TypeToken<ActionReceived>(){}.getType();

    public static Boolean removeAction(Context context, Integer id) {
        return shared.remove(context, Definitions.SHARED_ACTIONS, id.toString());
    }

    public static List<ActionReceived> listActions(Context context) {
        return shared.getAllObjects(context, typeToken, Definitions.SHARED_ACTIONS);
    }

    public static void saveAction(Context context, ActionReceived received) {
        shared.set(context, Definitions.SHARED_ACTIONS, received.id.toString(), received);
    }

    public static ActionReceived getActionByKey(Context context, Integer id){
        return shared.get(context, typeToken, Definitions.SHARED_ACTIONS, id.toString());
    }

    public static void cancelAllActions(Context context) {
        List<ActionReceived> receivedList = shared.getAllObjects(context, typeToken, Definitions.SHARED_ACTIONS);
        if (receivedList != null){
            for (ActionReceived received : receivedList) {
                cancelAction(context, received.id);
            }
        }
    }

    public static void cancelAction(Context context, Integer id) {
        ActionReceived received = getActionByKey(context, id);
        if(received !=null)
            removeAction(context, received.id);
    }
}